src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -1143,6 +1143,9 @@ abstract class ClassfileParser {
         (newStub(name.toTypeName), newStub(name.toTermName))
       } else {
         val cls = owner.newClass(name.toTypeName, NoPosition, sflags)
+        if (cls.isProtected) {
+          cls.privateWithin = cls.enclosingPackage
+        }
```

Comment:
This looks good, I think it needs to be done for the `mod` below too. A single check `if (jflags.isProtected)` should be enough.